fix(falcon.routing.DefaultRouter): Do not mask variable path segments

kgriffs opened the pull request:
fix(falcon.routing.DefaultRouter): Do not mask variable path segments

When a single segment in a requested path can match more than one node at that branch in the routing tree, and the first branch taken happens to be the wrong one (i.e., the subsequent nodes do not match, but they would have under a different branch), do not mask the other branches that could result in a successful resolution of the requested path.

I chose to implement this by not immediately returning None when a branch prediction turns out to be false. Alternatively, I could have introduced functions but that would likely incur more overhead vs. performing a few additional node comparisons on the way back up the routing tree. I did, however, include a simple optimization that allows a "fast return" when only simple nodes are involved in the branching, since in that case there is only one possible route.

More advanced optimizations are possible, and may be attempted in the future. For example:

* Graft the alternate branches directly into each point where the
  routing tree would otherwise 'return None'. This would balloon the
  size of the generated code, but would be akin to inline expansion
  of a routing representing the traversal of each branch where
  branch mispredictions are possible.
* Introspect competing regular expressions to determine whether it
  is actually possible for them to match the same string given in
  the path.
* Set a flag that can be referenced when returning from a branch
  misprediction in order to short-circuit checks. This may or may
  not be faster that some of the checks (i.e., those that check
  path length would not benefit from this flag, while those that
  index into the path list and perform a string comparison may be).

Fixes #702

kgriffs commented:
Routing tree from the test: https://github.com/falconry/falcon/pull/723

jmvrbanac commented:
Interesting. Seems to look good to me.

kgriffs commented:
Sorry, I had the wrong link for the routing tree. Updated.

fxfitz commented on falcon/routing/compiled.py:
```diff
@@ -119,6 +119,18 @@ def line(text, indent_offset=0):
+                var_nodes = [node for node in nodes if node.is_var]
+                found_var_nodes = bool(var_nodes)
+
+                fast_return = not found_var_nodes
```
This section all looks fine to me.

fxfitz commented on falcon/routing/compiled.py:
```diff
@@ -162,10 +174,11 @@ def line(text, indent_offset=0):
             if node.resource is None:
-                line('return None')
+                if fast_return:
+                    line('return None')
```
One thing that isn't overly clear to me is why we're adding these lines only if fast-return is True?  What happens If fast-return is false, such as the case where there are variable nodes?

(this is for all the `line('return None')` lines)

fxfitz replied:
Hmm. I think I understand it now after going through it once more.

:+1:

fxfitz commented on tests/test_default_router.py:
```diff
@@ -167,6 +178,32 @@ def test_variable(self):
+        self.assertEqual(resource.resource_id, 21)
+
+        resource, method_map, params = self.router.find('/gists/1776/pdf')
+        self.assertEqual(resource.resource_id, 21)
```
Can we add test assertions for when a router *can't* find a route, instead of just for ones where there is a clear route defined?

jmvrbanac replied:
That would probably be a good idea

jmvrbanac commented:
Outside of of @fxfitz's suggestion, :+1:

kgriffs commented:
@fxfitz, @jmvrbanac Thanks for the feedback! I rebased and added several negative tests, as well as adding tests for a couple of the specific cases that were reported in #702.

jmvrbanac commented:
:+1:

fxfitz commented:
LGTM too!
